// @parseaple/bplist/dist/index.mjs
import fs from "node:fs";

//#region src/index.ts
let maxObjectSize = 100 * 1e3 * 1e3;
let maxObjectCount = 32768;
const EPOCH = 9783072e5;
var UID = class {
	UID;
	constructor(id) {
		this.UID = id;
	}
};
const parseFile = (fileNameOrBuffer, callback) => {
	return new Promise((resolve, reject) => {
		const tryParseBuffer = (buffer) => {
			let err = null;
			let result;
			try {
				result = parseBuffer(buffer);
				resolve(result);
			} catch (ex) {
				err = ex;
				reject(err);
			} finally {
				if (callback) callback(err, result);
			}
		};
		if (Buffer.isBuffer(fileNameOrBuffer)) return tryParseBuffer(fileNameOrBuffer);
		fs.readFile(fileNameOrBuffer, (err, data) => {
			if (err) {
				reject(err);
				if (callback) callback(err);
				return;
			}
			tryParseBuffer(data);
		});
	});
};
const parseFileSync = (fileNameOrBuffer) => {
	let buffer;
	if (!Buffer.isBuffer(fileNameOrBuffer)) buffer = fs.readFileSync(fileNameOrBuffer);
	else buffer = fileNameOrBuffer;
	return parseBuffer(buffer);
};
const parseBuffer = (buffer) => {
	if (buffer.subarray(0, 6).toString("utf8") !== "bplist") throw new Error("Invalid binary plist. Expected 'bplist' at offset 0.");
	const trailer = buffer.subarray(buffer.length - 32, buffer.length);
	const offsetSize = trailer.readUInt8(6);
	const objectRefSize = trailer.readUInt8(7);
	const numObjects = readUInt64BE(trailer, 8);
	const topObject = readUInt64BE(trailer, 16);
	const offsetTableOffset = readUInt64BE(trailer, 24);
	if (numObjects > maxObjectCount) throw new Error("maxObjectCount exceeded");
	const offsetTable = [];
	for (let i = 0; i < numObjects; i++) offsetTable[i] = readUInt(buffer.subarray(offsetTableOffset + i * offsetSize, offsetTableOffset + (i + 1) * offsetSize), 0);
	const parseObject = (tableOffset) => {
		const offset = offsetTable[tableOffset];
		const type = buffer[offset];
		const objType = (type & 240) >> 4;
		const objInfo = type & 15;
		switch (objType) {
			case 0: return parseSimple();
			case 1: return parseInteger();
			case 8: return parseUID();
			case 2: return parseReal();
			case 3: return parseDate();
			case 4: return parseData();
			case 5: return parsePlistString();
			case 6: return parsePlistString(true);
			case 10: return parseArray();
			case 13: return parseDictionary();
			default: throw new Error(`Unhandled type 0x${objType.toString(16)}`);
		}
		function parseSimple() {
			switch (objInfo) {
				case 0: return null;
				case 8: return false;
				case 9: return true;
				case 15: return null;
				default: throw new Error(`Unhandled simple type 0x${objType.toString(16)}`);
			}
		}
		function bufferToHexString(buffer) {
			let str = "";
			let i;
			for (i = 0; i < buffer.length; i++) if (buffer[i] !== 0) break;
			for (; i < buffer.length; i++) {
				const part = `00${buffer[i].toString(16)}`;
				str += part.substring(part.length - 2);
			}
			return str;
		}
		function parseInteger() {
			const length = 2 ** objInfo;
			if (length < maxObjectSize) {
				const data = buffer.subarray(offset + 1, offset + 1 + length);
				if (length === 16) {
					const str = bufferToHexString(data);
					return BigInt(`0x${str}`);
				}
				return data.reduce((acc, curr) => {
					acc <<= 8;
					acc |= curr & 255;
					return acc;
				}, 0);
			}
			throw new Error(`Too little heap space available! Wanted to read ${length} bytes, but only ${maxObjectSize} are available.`);
		}
		function parseUID() {
			const length = objInfo + 1;
			if (length < maxObjectSize) return new UID(readUInt(buffer.subarray(offset + 1, offset + 1 + length)));
			throw new Error(`Too little heap space available! Wanted to read ${length} bytes, but only ${maxObjectSize} are available.`);
		}
		function parseReal() {
			const length = 2 ** objInfo;
			if (length < maxObjectSize) {
				const realBuffer = buffer.subarray(offset + 1, offset + 1 + length);
				if (length === 4) return realBuffer.readFloatBE(0);
				if (length === 8) return realBuffer.readDoubleBE(0);
			}
			throw new Error(`Too little heap space available! Wanted to read ${length} bytes, but only ${maxObjectSize} are available.`);
		}
		function parseDate() {
			if (objInfo !== 3) console.error(`Unknown date type :${objInfo}. Parsing anyway...`);
			const dateBuffer = buffer.subarray(offset + 1, offset + 9);
			return new Date(EPOCH + 1e3 * dateBuffer.readDoubleBE(0));
		}
		function parseData() {
			let dataoffset = 1;
			let length = objInfo;
			if (objInfo === 15) {
				const int_type = buffer[offset + 1];
				const intType = (int_type & 240) / 16;
				if (intType !== 1) console.error(`0x4: UNEXPECTED LENGTH-INT TYPE! ${intType}`);
				const intLength = 2 ** (int_type & 15);
				dataoffset = 2 + intLength;
				length = readUInt(buffer.subarray(offset + 2, offset + 2 + intLength));
			}
			if (length < maxObjectSize) return buffer.subarray(offset + dataoffset, offset + dataoffset + length);
			throw new Error(`Too little heap space available! Wanted to read ${length} bytes, but only ${maxObjectSize} are available.`);
		}
		function parsePlistString(isUtf16 = false) {
			let enc = "utf8";
			let length = objInfo;
			let stroffset = 1;
			if (objInfo === 15) {
				const int_type = buffer[offset + 1];
				const intType = (int_type & 240) / 16;
				if (intType !== 1) console.error(`UNEXPECTED LENGTH-INT TYPE! ${intType}`);
				const intLength = 2 ** (int_type & 15);
				stroffset = 2 + intLength;
				length = readUInt(buffer.subarray(offset + 2, offset + 2 + intLength));
			}
			length *= isUtf16 ? 2 : 1;
			if (length < maxObjectSize) {
				let plistString = buffer.subarray(offset + stroffset, offset + stroffset + length);
				if (isUtf16) {
					plistString = swapBytes(plistString);
					enc = "ucs2";
				}
				return plistString.toString(enc);
			}
			throw new Error(`Too little heap space available! Wanted to read ${length} bytes, but only ${maxObjectSize} are available.`);
		}
		function parseArray() {
			let length = objInfo;
			let arrayoffset = 1;
			if (objInfo === 15) {
				const int_type = buffer[offset + 1];
				const intType = (int_type & 240) / 16;
				if (intType !== 1) console.error(`0xa: UNEXPECTED LENGTH-INT TYPE! ${intType}`);
				const intLength = 2 ** (int_type & 15);
				arrayoffset = 2 + intLength;
				length = readUInt(buffer.subarray(offset + 2, offset + 2 + intLength));
			}
			if (length * objectRefSize > maxObjectSize) throw new Error("Too little heap space available!");
			const array = [];
			for (let i = 0; i < length; i++) array[i] = parseObject(readUInt(buffer.subarray(offset + arrayoffset + i * objectRefSize, offset + arrayoffset + (i + 1) * objectRefSize)));
			return array;
		}
		function parseDictionary() {
			let length = objInfo;
			let dictoffset = 1;
			if (objInfo === 15) {
				const int_type = buffer[offset + 1];
				const intType = (int_type & 240) / 16;
				if (intType !== 1) console.error(`0xD: UNEXPECTED LENGTH-INT TYPE! ${intType}`);
				const intLength = 2 ** (int_type & 15);
				dictoffset = 2 + intLength;
				length = readUInt(buffer.subarray(offset + 2, offset + 2 + intLength));
			}
			if (length * 2 * objectRefSize > maxObjectSize) throw new Error("Too little heap space available!");
			const dict = {};
			for (let i = 0; i < length; i++) {
				const keyRef = readUInt(buffer.subarray(offset + dictoffset + i * objectRefSize, offset + dictoffset + (i + 1) * objectRefSize));
				const valRef = readUInt(buffer.subarray(offset + dictoffset + length * objectRefSize + i * objectRefSize, offset + dictoffset + length * objectRefSize + (i + 1) * objectRefSize));
				const key = parseObject(keyRef);
				dict[key] = parseObject(valRef);
			}
			return dict;
		}
	};
	return [parseObject(topObject)];
};
function readUInt(buffer, start = 0) {
	let l = 0;
	for (let i = start; i < buffer.length; i++) {
		l <<= 8;
		l |= buffer[i] & 255;
	}
	return l;
}
function readUInt64BE(buffer, start) {
	return buffer.subarray(start, start + 8).readUInt32BE(4);
}
function swapBytes(buffer) {
	const copy = Buffer.from(buffer);
	const len = copy.length;
	for (let i = 0; i < len; i += 2) {
		const a = copy[i];
		copy[i] = copy[i + 1];
		copy[i + 1] = a;
	}
	return copy;
}

//#endregion
export { UID, maxObjectCount, maxObjectSize, parseBuffer, parseFile, parseFileSync };
