// @parseaple/typedstream/dist/index.mjs
import { AssertionError } from "node:assert";
import { parseBuffer } from "@parseaple/bplist";
//#region src/encodings.ts
/**
* Find the end index of the encoding starting at index start.
*
* The encoding is not validated very extensively.
* There are no guarantees what happens for invalid encodings;
* an error may be raised,
* or a bogus end index may be returned.
* Callers are expected to check that the returned end index actually results in a valid encoding.
* @param encoding
* @param start
*/
function endOfEncoding(encoding, start) {
	if (start > encoding.length) throw new EvalError(`Start index ${start} not in range(${encoding.length})`);
	let parenDepth = 0;
	let i = start;
	while (i < encoding.length) {
		const c = encoding.substring(i, i + 1);
		if ([
			"(",
			"[",
			"{"
		].includes(c)) {
			parenDepth++;
			i++;
		} else if (parenDepth > 0) {
			if ([
				")",
				"]",
				"}"
			].includes(c)) parenDepth--;
			i++;
			if (parenDepth === 0) return i;
		} else return i + 1;
	}
	throw new EvalError("Incomplete encoding");
}
/**
* Split apart multiple type encodings contained in a single encoding string.
* @param encodings
*/
function* splitEncodings(encodings) {
	let start = 0;
	while (start < encodings.length) {
		const end = endOfEncoding(encodings, start);
		yield encodings.substring(start, end);
		start = end;
	}
}
/**
* Build an array type encoding from a length and an element type encoding.
*
* .. note::
*
* This function currently doesn't perform any checking on ``element_type_encoding``,
* but such checks may be added in the future.
* ``element_type_encoding`` should always be a valid type encoding string.
* @param arrayEncoding
*/
function parseArrayEncoding(arrayEncoding) {
	if (!arrayEncoding.startsWith("[")) throw new EvalError(`Missing opening bracket in array type encoding ${arrayEncoding}`);
	if (!arrayEncoding.endsWith("]")) throw new EvalError(`Missing closing bracket in array type encoding ${arrayEncoding}`);
	let i = 1;
	while (i < arrayEncoding.length - 1) {
		if (![
			"0",
			"1",
			"2",
			"3",
			"4",
			"5",
			"6",
			"7",
			"8",
			"9"
		].includes(arrayEncoding.charAt(i))) break;
		i++;
	}
	const lengthString = arrayEncoding.substring(1, i);
	const elementTypeEncoding = arrayEncoding.substring(i, arrayEncoding.length - 1);
	return {
		length: parseInt(lengthString, 10),
		elementTypeEncoding
	};
}
function buildArrayEncoding(length, elementTypeEncoding) {
	if (length < 0) throw new EvalError(`Array length cannot be negative: ${length}`);
	return `[${length.toString()}${elementTypeEncoding}]`;
}
/**
* Parse an array type encoding into its name and field type encodings.
* @param structEncoding
*/
function parseStructEncoding(structEncoding) {
	if (!structEncoding.startsWith("{")) throw new EvalError(`Missing opening brace in struct type encoding ${structEncoding}`);
	if (!structEncoding.endsWith("}")) throw new EvalError(`Missing closing brace in struct type encoding ${structEncoding}`);
	const firstNestedBrace = structEncoding.indexOf("{", 1);
	const nameSearchEnd = firstNestedBrace === -1 ? structEncoding.length - 1 : firstNestedBrace;
	const equalsInRange = structEncoding.substring(1, nameSearchEnd).indexOf("=");
	const equalsPos = equalsInRange === -1 ? -1 : equalsInRange + 1;
	let name;
	let fieldTypeEncodingString;
	if (equalsPos === -1) fieldTypeEncodingString = structEncoding.substring(1, structEncoding.length - 1);
	else {
		name = structEncoding.substring(1, equalsPos);
		fieldTypeEncodingString = structEncoding.substring(equalsPos + 1, structEncoding.length - 1);
	}
	const fieldTypeEncodings = [...splitEncodings(fieldTypeEncodingString)];
	return {
		name,
		fieldTypeEncodings
	};
}
function encodingMatchesExpected(actualEncoding, expectedEncoding) {
	if (actualEncoding.startsWith("{") && expectedEncoding.startsWith("{")) {
		const { name: actualName, fieldTypeEncodings: actualFieldTypeEncodings } = parseStructEncoding(actualEncoding);
		const { name: expectedName, fieldTypeEncodings: expectedFieldTypeEncodings } = parseStructEncoding(expectedEncoding);
		return (actualName === null || actualName === "?" || actualName === expectedName) && allEncodingsMatchExpected(actualFieldTypeEncodings, expectedFieldTypeEncodings);
	} else if (actualEncoding.startsWith("[") && expectedEncoding.startsWith("[")) {
		const { length: actualLength, elementTypeEncoding: actualElementTypeEncoding } = parseArrayEncoding(actualEncoding);
		const { length: expectedLength, elementTypeEncoding: expectedElementTypeEncoding } = parseArrayEncoding(expectedEncoding);
		return actualLength === expectedLength && encodingMatchesExpected(actualElementTypeEncoding, expectedElementTypeEncoding);
	} else return actualEncoding === expectedEncoding;
}
function allEncodingsMatchExpected(actualEncodings, expectedEncodings) {
	if (actualEncodings.length !== expectedEncodings.length) return false;
	for (let i = 0; i < actualEncodings.length; i++) if (!encodingMatchesExpected(actualEncodings[i], expectedEncodings[i])) return false;
	return true;
}
//#endregion
//#region src/stream.ts
const STREAMER_VERSION_OLD_NEXTSTEP = 3;
const STREAMER_VERSION_CURRENT = 4;
const SIGNATURE_LENGTH = 11;
const SIGNATURE_TO_BYTE_ORDER_MAP = {
	typedstream: "BE",
	streamtyped: "LE"
};
const TAG_INTEGER_2 = -127;
const TAG_INTEGER_4 = -126;
const TAG_FLOATING_POINT = -125;
const TAG_NEW = -124;
const TAG_NIL = -123;
const TAG_END_OF_OBJECT = -122;
const FIRST_TAG = -128;
const LAST_TAG = -111;
function inTagRange(n) {
	if (n == null) return false;
	return n >= FIRST_TAG && n <= LAST_TAG;
}
const FIRST_REFERENCE_NUMBER = LAST_TAG + 1;
/**
* Decode a reference number (as stored in a typedstream) to a regular zero-based index.
*/
function decodeReferenceNumber(encoded) {
	return encoded - FIRST_REFERENCE_NUMBER;
}
/**
* Raised by :class:`TypedStreamReader` if the typedstream data is invalid or doesn't match the expected structure.
*/
var InvalidTypedStreamError = class extends Error {};
var EOFError = class extends Error {};
/**
* Marks the beginning of a group of values prefixed by a type encoding string.
*/
var BeginTypedValues = class {
	encodings;
	constructor(encodings) {
		this.encodings = encodings;
	}
};
/**
* Marks the end of a group of values prefixed by a type encoding string.
*
* This event is provided for convenience and doesn't correspond to any data in the typedstream.
*/
var EndTypedValues = class {};
/**
* A reference to a previously read object.
*/
var ObjectReference = class {
	referencedType;
	number;
	constructor(referencedType, number) {
		this.referencedType = referencedType;
		this.number = number;
	}
};
(function(_ObjectReference) {
	_ObjectReference.Type = /* @__PURE__ */ function(Type) {
		Type["C_STRING"] = "C string";
		Type["CLASS"] = "class";
		Type["OBJECT"] = "object";
		return Type;
	}({});
})(ObjectReference || (ObjectReference = {}));
/**
* A NeXTSTEP atom (NXAtom), i. e. a shared/deduplicated C string.
*
* In (Objective-)C on NeXTSTEP,
* atoms are immutable C strings that have been deduplicated so that two atoms with the same content always have the same address.
* This allows checking two atoms for equality by just comparing the pointers/addresses,
* instead of having to compare their contents.
* Other than that,
* atoms behave like regular C strings.
*
* Mac OS X/macOS no longer supports atoms and throws an exception when attempting to decode them using ``NSUnarchiver``.
*
* This is a thin wrapper around a plain :class:`bytes` object.
* The wrapper class is used to distinguish atoms from untyped bytes.
*/
var Atom = class {
	contents;
	constructor(contents) {
		this.contents = contents;
	}
};
/**
* An Objective-C selector.
*
* This is a thin wrapper around a plain :class:`bytes` object.
* The wrapper class is used to distinguish selector values from untyped bytes.
*/
var Selector = class {
	name;
	constructor(name) {
		this.name = name;
	}
};
/**
* Information about a C string as it is stored in a typedstream.
*
* This is a thin wrapper around a plain :class:`bytes` object.
* The wrapper class is used to distinguish typed C string values from untyped bytes.
*/
var CString = class {
	contents;
	constructor(contents) {
		this.contents = contents;
	}
};
/**
* Information about a class (name and version),
* stored literally in a chain of superclasses in a typedstream.
*
* A class in a typedstream can be stored literally, as a reference, or be ``Nil``.
* A literally stored class is always followed by information about its superclass.
* If the superclass information is also stored literally,
* it is again followed by information about its superclass.
* This chain continues until a class is reached that has been stored before
* (in which case it is stored as a reference)
* or a root class is reached
* (in which case the superclass is ``Nil``).
*
* The beginning and end of such a chain of superclasses are not marked explicitly in a typedstream,
* and no events are generated when a superclass chain begins or ends.
* A superclass chain begins implicitly when a literally stored class is encountered
* (if no chain is already in progress),
* and the chain ends after the first non-literal (i. e. reference or ``Nil``) class.
*/
var SingleClass = class {
	name;
	version;
	constructor(name, version) {
		this.name = name;
		this.version = version;
	}
};
/**
* Marks the beginning of a literally stored object.
*
* This event is followed by information about the object's class,
* stored as a chain of class information (see :class:`SingleClass`).
* This class chain is followed by an arbitrary number of type-prefixed value groups,
* which represent the object's contents.
* The object ends when an :class:`EndObject` is encountered where the next value group would start.
*/
var BeginObject = class {};
/**
* Marks the end of a literally stored object.
*/
var EndObject = class {};
/**
* Represents an array of bytes (signed or unsigned char).
*
* For performance and simplicity,
* such arrays are read all at once and represented as a single event,
* instead of generating one event per element as for other array element types.
*/
var ByteArray = class {
	elementEncoding;
	data;
	constructor(elementEncoding, data) {
		this.elementEncoding = elementEncoding;
		this.data = data;
	}
};
/**
* Marks the beginning of an array.
*
* This event is provided for convenience and doesn't directly correspond to data in the typedstream.
* The array length and element type information provided in this event actually comes from the arrays's type encoding.
*
* This event is followed by the element values,
* which are not explicitly type-prefixed,
* as they all have the type specified in the array type encoding.
* The end of the array is not marked in the typedstream data,
* as it can be determined based on the length and element type,
* but for convenience,
* an :class:`EndArray` element is generated after the last array element.
*
* This event is *not* generated for arrays of bytes (signed or unsigned char) -
* such arrays are represented as single :class:`ByteArray` events instead.
*/
var BeginArray = class {
	elementEncoding;
	length;
	constructor(elementEncoding, length) {
		this.elementEncoding = elementEncoding;
		this.length = length;
	}
};
/**
* Marks the end of an array.
*
* This event is provided for convenience and doesn't correspond to any data in the typedstream.
*/
var EndArray = class {};
/**
* Marks the beginning of a struct.
*
* This event is provided for convenience and doesn't directly correspond to data in the typedstream.
* The struct name and field type information provided in this event actually comes from the struct's type encoding.
*
* This event is followed by the field values,
* which are not explicitly type-prefixed (unlike in objects),
* as their types are specified in the struct type encoding.
* The end of the struct is not marked in the typedstream data,
* as it can be determined based on the type information,
* but for convenience,
* an :class:`EndStruct` element is generated after the last struct field.
*/
var BeginStruct = class {
	name;
	fieldEncodings;
	constructor(fieldEncodings, name) {
		this.name = name;
		this.fieldEncodings = fieldEncodings;
	}
};
/**
* Marks the end of a struct.
*
* This event is provided for convenience and doesn't correspond to any data in the typedstream.
*/
var EndStruct = class {};
/**
* Reads typedstream data from a raw byte stream.
*/
var TypedStreamReader = class {
	EOF_MESSAGE = "End of typedstream reached";
	data;
	pos = 0;
	sharedStringTable;
	streamerVersion;
	byteOrder;
	systemVersion;
	eventsIterator;
	/**
	* Create a :class:`TypedStreamReader` that reads data from the given raw byte stream.
	*
	* @param data The raw byte stream from which to read the typedstream data.
	* By default this is ``False`` and callers are expected to close the raw stream themselves after closing the :class:`TypedStreamReader`.
	*/
	constructor(data) {
		this.data = data;
		this.sharedStringTable = [];
		this.streamerVersion = this.readInteger(false);
		const signatureLength = this.readInteger(false);
		if (this.streamerVersion < STREAMER_VERSION_OLD_NEXTSTEP || this.streamerVersion > STREAMER_VERSION_CURRENT) throw new InvalidTypedStreamError(`Invalid streamer version: ${this.streamerVersion}`);
		else if (this.streamerVersion === STREAMER_VERSION_OLD_NEXTSTEP) throw new InvalidTypedStreamError(`Old NeXTSTEP streamer version (${this.streamerVersion}) not supported (yet?)`);
		if (signatureLength !== SIGNATURE_LENGTH) throw new InvalidTypedStreamError(`The signature string must be exactly ${SIGNATURE_LENGTH} bytes long, not ${signatureLength}`);
		const signature = this.readExact(signatureLength).toString("ascii");
		this.byteOrder = SIGNATURE_TO_BYTE_ORDER_MAP[signature];
		this.systemVersion = this.readInteger(false);
		this.eventsIterator = this.readAllValues();
	}
	next() {
		return this.eventsIterator.next();
	}
	[Symbol.iterator]() {
		return this.eventsIterator;
	}
	/**
	* Read byte_count bytes from the raw stream and raise an exception if too few bytes are read
	* (i. e. if EOF was hit prematurely).
	*/
	readExact(byteCount) {
		const end = this.pos + byteCount;
		const result = this.data.subarray(this.pos, end);
		this.pos = end;
		return result;
	}
	/**
	* Read a head byte.
	*
	* @param head If ``None``, the head byte is read normally from the stream.
	* Otherwise, the passed-in head byte is returned and no read is performed.
	* This parameter is provided to simplify a common pattern in this class's internal methods,
	* where methods that need to read a head byte
	* can alternatively accept an already read head byte as a parameter
	* and skip the read operation.
	* This mechanism is used to allow a limited form of lookahead for the head byte,
	* which is needed to parse string and object references and to detect end-of-object markers.
	* @return The read or passed in head byte.
	* @private
	*/
	readHeadByte(head) {
		if (head == null) head = this.readExact(1).readInt8();
		return head;
	}
	/**
	* Read a low-level integer value.
	*
	* @param signed Whether to treat the integer as signed or unsigned.
	* @param head An already read head byte to use, or ``None`` if the head byte should be read from the stream.
	* @return The decoded integer value.
	* @private
	*/
	readInteger(signed, head) {
		head = this.readHeadByte(head);
		if (!inTagRange(head)) if (signed) return head;
		else return head & 255;
		else if (head === TAG_INTEGER_2) if (this.byteOrder === "LE") if (signed) return this.readExact(2).readInt16LE();
		else return this.readExact(2).readUint16LE();
		else if (signed) return this.readExact(2).readInt16BE();
		else return this.readExact(2).readUint16BE();
		else if (head === TAG_INTEGER_4) if (this.byteOrder === "LE") if (signed) return this.readExact(4).readInt32LE();
		else return this.readExact(4).readUint32LE();
		else if (signed) return this.readExact(4).readInt32BE();
		else return this.readExact(4).readUint32BE();
		else throw new InvalidTypedStreamError(`Invalid head tag in this context: ${head} (${head & 255}`);
	}
	/**
	* Read a low-level single-precision float value.
	*
	* @param head An already read head byte to use, or ``None`` if the head byte should be read from the stream.
	* @return The decoded float value
	* @private
	*/
	readFloat(head) {
		head = this.readHeadByte(head);
		if (head === TAG_FLOATING_POINT) if (this.byteOrder === "LE") return this.readExact(4).readFloatLE();
		else return this.readExact(4).readFloatBE();
		else return this.readInteger(true, head);
	}
	/**
	* Read a low-level double-precision float value.
	*
	* @param head An already read head byte to use, or ``None`` if the head byte should be read from the stream.
	* @return The decoded double value
	* @private
	*/
	readDouble(head) {
		head = this.readHeadByte(head);
		if (head === TAG_FLOATING_POINT) if (this.byteOrder === "LE") return this.readExact(8).readDoubleLE();
		else return this.readExact(8).readDoubleBE();
		else return this.readInteger(true, head);
	}
	/**
	* Read a low-level string value.
	*
	* Strings in typedstreams have no specificed encoding,
	* so the string data is returned as raw :class:`bytes`.
	* (In practice, they usually consist of printable ASCII characters.)
	*
	* @param head An already read head byte to use, or ``None`` if the head byte should be read from the stream.
	* @return The read string data, which may be ``nil``/``None``.
	* @private
	*/
	readUnsharedString(head) {
		head = this.readHeadByte(head);
		if (head === TAG_NIL) return;
		const length = this.readInteger(false, head);
		return this.readExact(length);
	}
	/**
	* Read a low-level shared string value.
	*
	* A shared string value may either be stored literally (as an unshared string)
	* or as a reference to a previous literally stored shared string.
	* Literal shared strings are appended to the :attr:`shared_string_table` after they are read,
	* so that they can be referenced by later non-literal shared strings.
	* This happens transparently to the caller -
	* in both cases the actual string data is returned.
	*
	* @param head An already read head byte to use, or ``None`` if the head byte should be read from the stream.
	* @return The read string data, which may be ``nil``/``None``.
	* @private
	*/
	readSharedString(head) {
		head = this.readHeadByte(head);
		if (head === TAG_NIL) return;
		else if (head === TAG_NEW) {
			const string = this.readUnsharedString();
			if (string == null) throw new InvalidTypedStreamError("Literal shared string cannot contain a nil unshared string");
			this.sharedStringTable.push(string);
			return string;
		} else {
			const decoded = decodeReferenceNumber(this.readInteger(true, head));
			return this.sharedStringTable[decoded];
		}
	}
	/**
	* Read an object reference value.
	*
	* Despite the name,
	* object references can't just refer to objects,
	* but also to classes or C strings.
	* The type of object that a reference refers to is always clear from context
	* and is not explicitly stored in the typedstream.
	*
	* @param referencedType The type of object that the reference refers to.
	* @param head An already read head byte to use, or ``None`` if the head byte should be read from the stream.
	* @return The read object reference.
	* @private
	*/
	readObjectReference(referencedType, head) {
		const referenceNumber = this.readInteger(true, head);
		return new ObjectReference(referencedType, decodeReferenceNumber(referenceNumber));
	}
	/**
	* Read a C string value.
	*
	* A C string value may either be stored literally
	* or as a reference to a previous literally stored C string value.
	* Literal C string values are returned as :class:`CString` objects.
	* C string values stored as references are returned as :class:`ObjectReference` objects
	* and are not automatically dereferenced.
	*
	* @param head An already read head byte to use, or ``None`` if the head byte should be read from the stream.
	* @return The read C string value or reference, which may be ``nil``/``None``.
	* @private
	*/
	readCString(head) {
		head = this.readHeadByte(head);
		if (head === TAG_NIL) return;
		else if (head === TAG_NEW) {
			const string = this.readSharedString();
			if (string == null) throw new InvalidTypedStreamError("Literal C string cannot contain a nil shared string");
			return new CString(string);
		} else return this.readObjectReference(ObjectReference.Type.C_STRING, head);
	}
	/**
	* Iteratively read a class object from the typedstream.
	*
	* @param head An already read head byte to use, or ``None`` if the head byte should be read from the stream.
	* @return An iterable of events representing the class object.
	* See :class:`SingleClass` for information about what events are generated when and what they mean.
	* @private
	*/
	*readClass(head) {
		head = this.readHeadByte(head);
		while (head === TAG_NEW) {
			const name = this.readSharedString();
			if (name == null) throw new InvalidTypedStreamError("Class name cannot be nil");
			const version = this.readInteger(true);
			yield new SingleClass(name.toString("ascii"), version);
			head = this.readHeadByte();
		}
		if (head === TAG_NIL) yield void 0;
		else yield this.readObjectReference(ObjectReference.Type.CLASS, head);
	}
	/**
	* Iteratively read an object from the typedstream,
	* including all of its contents and the end of object marker.
	*
	* @param head An already read head byte to use, or ``None`` if the head byte should be read from the stream.
	* @return An iterable of events representing the object.
	* See :class:`BeginObject` and :class:`EndObject` for information about what events are generated when and what they mean.
	* @private
	*/
	*readObject(head) {
		head = this.readHeadByte(head);
		if (head === TAG_NIL) yield void 0;
		else if (head === TAG_NEW) {
			yield new BeginObject();
			yield* this.readClass();
			let nextHead = this.readHeadByte();
			while (nextHead !== TAG_END_OF_OBJECT) {
				yield* this.readTypedValues(nextHead);
				nextHead = this.readHeadByte();
			}
			yield new EndObject();
		} else yield this.readObjectReference(ObjectReference.Type.OBJECT, head);
	}
	/**
	* Iteratively read a single value with the type indicated by the given type encoding.
	*
	* The type encoding string must contain exactly one type
	* (although it may be a compound type like a struct or array).
	* Type encoding strings that might contain more than one value must first be split using :func:`_split_encodings`.
	*
	* @param typeEncoding
	* @param head An already read head byte to use, or ``None`` if the head byte should be read from the stream.
	* @return An iterable of events representing the object.
	* Simple values are represented by single events,
	* but more complex values (classes, objects, arrays, structs) usually generate multiple events.
	* @private
	*/
	*readValueWithEncoding(typeEncoding, head) {
		if (typeEncoding === "C" || typeEncoding === "B") yield this.readExact(1).readUInt8();
		else if (typeEncoding === "c") yield this.readExact(1).readInt8();
		else if ([
			"S",
			"I",
			"L",
			"Q"
		].includes(typeEncoding)) yield this.readInteger(false, head);
		else if ([
			"s",
			"i",
			"l",
			"q"
		].includes(typeEncoding)) yield this.readInteger(true, head);
		else if (typeEncoding === "f") yield this.readFloat(head);
		else if (typeEncoding === "d") yield this.readDouble(head);
		else if (typeEncoding === "*") yield this.readCString(head);
		else if (typeEncoding === "%") yield new Atom(this.readSharedString(head));
		else if (typeEncoding === ":") yield new Selector(this.readSharedString(head));
		else if (typeEncoding === "+") yield this.readUnsharedString(head);
		else if (typeEncoding === "#") yield* this.readClass(head);
		else if (typeEncoding === "@") yield* this.readObject(head);
		else if (typeEncoding === "!") yield void 0;
		else if (typeEncoding.startsWith("[")) {
			const { length, elementTypeEncoding } = parseArrayEncoding(typeEncoding);
			if (["c", "C"].includes(elementTypeEncoding)) yield new ByteArray(elementTypeEncoding, this.readExact(length));
			else {
				yield new BeginArray(elementTypeEncoding, length);
				for (let i = 0; i < length; i++) yield* this.readValueWithEncoding(elementTypeEncoding);
				yield new EndArray();
			}
		} else if (typeEncoding.startsWith("{")) {
			const { name, fieldTypeEncodings } = parseStructEncoding(typeEncoding);
			yield new BeginStruct(fieldTypeEncodings, name);
			for (const fieldTypeEncoding of fieldTypeEncodings) yield* this.readValueWithEncoding(fieldTypeEncoding);
			yield new EndStruct();
		} else throw new InvalidTypedStreamError(`Don't know how to read a value with type encoding ${typeEncoding}`);
	}
	/**
	* Iteratively read the next group of typed values from the stream.
	*
	* The type encoding string is decoded to determine the type of the following values.
	*
	* @param head An already read head byte to use, or ``None`` if the head byte should be read from the stream.
	* @param endOfStreamOk Whether reaching the end of the data stream is an acceptable condition.
	* If this method is called when the end of the stream is reached,
	* an :class:`EOFError` is raised if this parameter is true,
	* and an :class:`InvalidTypedStreamError` is raised if it is false.
	* If the end of the stream is reached in the middle of reading a value
	* (not right at the beginning),
	* the exception is always an :class:`InvalidTypedStreamError`,
	* regardless of the value of this parameter.
	* @return An iterable of events representing the typed values.
	* See :class:`BeginTypedValues` and :class:`EndTypedValues` for information about what events are generated when and what they mean.
	*/
	*readTypedValues(head, endOfStreamOk = false) {
		try {
			head = this.readHeadByte(head);
		} catch (e) {
			if (e instanceof RangeError && e.message === "Attempt to access memory outside buffer bounds" && endOfStreamOk) throw new EOFError(this.EOF_MESSAGE);
			else throw e;
		}
		const encodingString = this.readSharedString(head)?.toString("ascii");
		if (encodingString == null) throw new InvalidTypedStreamError("Encountered nil type encoding string");
		const typeEncodings = Array.from(splitEncodings(encodingString));
		yield new BeginTypedValues(typeEncodings);
		for (const typeEncoding of typeEncodings) yield* this.readValueWithEncoding(typeEncoding);
		yield new EndTypedValues();
	}
	/**
	* Iteratively read all values in the typedstream.
	*
	* @return An iterable of events representing the contents of the typedstream.
	* Top-level values in a typedstream are always prefixed with a type encoding.
	* See :class:`BeginTypedValues` and :class:`EndTypedValues` for information about what events are generated when and what they mean.
	* @private
	*/
	*readAllValues() {
		while (true) try {
			yield* this.readTypedValues(void 0, true);
		} catch (e) {
			if (e instanceof EOFError) return;
			else throw e;
		}
	}
};
//#endregion
//#region src/types/known_types.ts
var KnownArchivedObject = class {
	static archivedName;
	static consumeEnd;
	static initFromUnarchiver(_unarchiver, _archivedClass) {
		throw new Error("Method not implemented");
	}
};
const archivedClassesByName = /* @__PURE__ */ new Map();
function archivedClass(archivedName, consumeEnd = false) {
	return (target) => {
		target.archivedName = archivedName;
		target.consumeEnd = consumeEnd;
		archivedClassesByName.set(archivedName, target);
	};
}
const structClassesByEncoding = /* @__PURE__ */ new Map();
//#endregion
//#region \0@oxc-project+runtime@0.126.0/helpers/decorate.js
function __decorate(decorators, target, key, desc) {
	var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	return c > 3 && r && Object.defineProperty(target, key, r), r;
}
//#endregion
//#region src/types/foundation.ts
var _NSObject, _NSData, _NSMutableData, _NSDate, _NSString, _NSMutableString, _NSAttributedString, _NSMutableAttributedString, _NSValue, _NSNumber, _NSArray, _NSMutableArray, _NSSet, _NSMutableSet, _NSDictionary, _NSMutableDictionary;
let NSObject = _NSObject = class NSObject extends KnownArchivedObject {
	static initFromUnarchiver(_unarchiver, archivedClass) {
		if (archivedClass.version !== 0) throw new EvalError(`Unsupported version: ${archivedClass.version}`);
		return new _NSObject();
	}
};
NSObject = _NSObject = __decorate([archivedClass("NSObject")], NSObject);
let NSData = _NSData = class NSData extends NSObject {
	data;
	constructor(data) {
		super();
		this.data = data;
	}
	static initFromUnarchiver(unarchiver, archivedClass) {
		if (archivedClass.version !== 0) throw new EvalError(`Unsupported version: ${archivedClass.version}`);
		return new _NSData(Buffer.from(unarchiver.decodeDataObject()));
	}
};
NSData = _NSData = __decorate([archivedClass("NSData")], NSData);
let NSMutableData = _NSMutableData = class NSMutableData extends NSData {
	static initFromUnarchiver(unarchiver, archivedClass) {
		if (archivedClass.version !== 0) throw new EvalError(`Unsupported version: ${archivedClass.version}`);
		return new _NSMutableData(Buffer.from(unarchiver.decodeDataObject()));
	}
};
NSMutableData = _NSMutableData = __decorate([archivedClass("NSMutableData")], NSMutableData);
let NSDate = class NSDate extends NSObject {
	static {
		_NSDate = this;
	}
	static ABSOLUTE_REFERENCE_DATA = Date.UTC(2001, 1, 1);
	absoluteReferenceDateOffset;
	get value() {
		return new Date(_NSDate.ABSOLUTE_REFERENCE_DATA + 1e3 * this.absoluteReferenceDateOffset);
	}
	constructor(absoluteReferenceDateOffset) {
		super();
		this.absoluteReferenceDateOffset = absoluteReferenceDateOffset;
	}
	static initFromUnarchiver(unarchiver, archivedClass) {
		if (archivedClass.version !== 0) throw new EvalError(`Unsupported version: ${archivedClass.version}`);
		return new _NSDate(unarchiver.decodeValueOfType("d"));
	}
};
NSDate = _NSDate = __decorate([archivedClass("NSDate")], NSDate);
let NSString = _NSString = class NSString extends NSObject {
	string;
	constructor(value) {
		super();
		this.string = value;
	}
	static initFromUnarchiver(unarchiver, archivedClass) {
		if (archivedClass.version !== 1) throw new EvalError(`Unsupported version: ${archivedClass.version}`);
		return new _NSString(unarchiver.decodeValueOfType("+").toString("utf8"));
	}
};
NSString = _NSString = __decorate([archivedClass("NSString")], NSString);
let NSMutableString = _NSMutableString = class NSMutableString extends NSString {
	static initFromUnarchiver(unarchiver, archivedClass) {
		if (archivedClass.version !== 1) throw new EvalError(`Unsupported version: ${archivedClass.version}`);
		return new _NSMutableString(unarchiver.decodeValueOfType("+").toString("utf8"));
	}
};
NSMutableString = _NSMutableString = __decorate([archivedClass("NSMutableString")], NSMutableString);
let NSAttributedString = _NSAttributedString = class NSAttributedString extends NSString {
	runs;
	constructor(value, runs) {
		super(value);
		this.runs = runs;
	}
	static readString(unarchiver) {
		return unarchiver.decodeTypedValues().values[0].string;
	}
	static readRange(unarchiver) {
		const range = unarchiver.decodeTypedValues();
		return {
			reference: range.values[0],
			length: range.values[1]
		};
	}
	static readAttributeValue(unarchiver) {
		const dict = unarchiver.decodeTypedValues().values[0].contents;
		const attributeValue = {};
		for (const [key, value] of dict.entries()) if (value instanceof NSData) {
			if (unarchiver.binaryDecoding !== Unarchiver.BinaryDecoding.none) {
				const data = value.data;
				try {
					const decoded = Unarchiver.open(data).decodeAll();
					if (decoded.length === 1) attributeValue[key.string] = decoded[0];
					else attributeValue[key.string] = decoded;
				} catch {
					if (unarchiver.binaryDecoding === Unarchiver.BinaryDecoding.all) attributeValue[key.string] = data;
				}
			}
		} else if (value instanceof NSString) attributeValue[key.string] = value.string;
		else attributeValue[key.string] = value.value;
		return attributeValue;
	}
	static readAttributes(unarchiver, length) {
		const runs = [];
		let index = 0;
		const sharedAttributeValues = {};
		while (index < length) {
			const range = _NSAttributedString.readRange(unarchiver);
			if (!(range.reference in sharedAttributeValues)) sharedAttributeValues[range.reference] = _NSAttributedString.readAttributeValue(unarchiver);
			const attributeValue = sharedAttributeValues[range.reference];
			runs.push({
				range: [index, range.length],
				attributes: attributeValue
			});
			index += range.length;
		}
		return runs;
	}
	static initFromUnarchiver(unarchiver, archivedClass) {
		if (archivedClass.version !== 0) throw new EvalError(`Unsupported version: ${archivedClass.version}`);
		const value = _NSAttributedString.readString(unarchiver);
		return new _NSAttributedString(value, _NSAttributedString.readAttributes(unarchiver, value.length));
	}
};
NSAttributedString = _NSAttributedString = __decorate([archivedClass("NSAttributedString")], NSAttributedString);
let NSMutableAttributedString = _NSMutableAttributedString = class NSMutableAttributedString extends NSAttributedString {
	static initFromUnarchiver(unarchiver, archivedClass) {
		if (archivedClass.version !== 0) throw new EvalError(`Unsupported version: ${archivedClass.version}`);
		const value = _NSMutableAttributedString.readString(unarchiver);
		return new _NSMutableAttributedString(value, _NSMutableAttributedString.readAttributes(unarchiver, value.length));
	}
};
NSMutableAttributedString = _NSMutableAttributedString = __decorate([archivedClass("NSMutableAttributedString")], NSMutableAttributedString);
let NSValue = _NSValue = class NSValue extends NSObject {
	typeEncoding;
	value;
	constructor(typeEncoding, value) {
		super();
		this.typeEncoding = typeEncoding;
		this.value = value;
	}
	static initFromUnarchiver(unarchiver, archivedClass) {
		if (archivedClass.version !== 0) throw new EvalError(`Unsupported version: ${archivedClass.version}`);
		const typeEncoding = unarchiver.decodeValueOfType("*");
		const value = unarchiver.decodeValueOfType(typeEncoding);
		return new _NSValue(typeEncoding, value);
	}
};
NSValue = _NSValue = __decorate([archivedClass("NSValue")], NSValue);
let NSNumber = _NSNumber = class NSNumber extends NSValue {
	static initFromUnarchiver(unarchiver, archivedClass) {
		if (archivedClass.version !== 0) throw new EvalError(`Unsupported version: ${archivedClass.version}`);
		const typeEncoding = unarchiver.decodeValueOfType("*");
		const value = unarchiver.decodeValueOfType(typeEncoding);
		return new _NSNumber(typeEncoding, value);
	}
};
NSNumber = _NSNumber = __decorate([archivedClass("NSNumber")], NSNumber);
let NSArray = _NSArray = class NSArray extends NSObject {
	elements;
	constructor(elements) {
		super();
		this.elements = elements;
	}
	static initFromUnarchiver(unarchiver, archivedClass) {
		if (archivedClass.version !== 0) throw new EvalError(`Unsupported version: ${archivedClass.version}`);
		const count = unarchiver.decodeValueOfType("i");
		if (count < 0) throw new EvalError(`NSArray element count cannot be negative: ${count}`);
		const elements = [];
		for (let i = 0; i < count; i++) elements.push(unarchiver.decodeValueOfType("@"));
		return new _NSArray(elements);
	}
};
NSArray = _NSArray = __decorate([archivedClass("NSArray")], NSArray);
let NSMutableArray = _NSMutableArray = class NSMutableArray extends NSArray {
	static initFromUnarchiver(unarchiver, archivedClass) {
		if (archivedClass.version !== 0) throw new EvalError(`Unsupported version: ${archivedClass.version}`);
		const count = unarchiver.decodeValueOfType("i");
		if (count < 0) throw new EvalError(`NSMutableArray element count cannot be negative: ${count}`);
		const elements = [];
		for (let i = 0; i < count; i++) elements.push(unarchiver.decodeValueOfType("@"));
		return new _NSMutableArray(elements);
	}
};
NSMutableArray = _NSMutableArray = __decorate([archivedClass("NSMutableArray")], NSMutableArray);
let NSSet = _NSSet = class NSSet extends NSObject {
	elements;
	constructor(elements) {
		super();
		this.elements = elements;
	}
	static initFromUnarchiver(unarchiver, archivedClass) {
		if (archivedClass.version !== 0) throw new EvalError(`Unsupported version: ${archivedClass.version}`);
		const count = unarchiver.decodeValueOfType("I");
		const elements = [];
		for (let i = 0; i < count; i++) elements.push(unarchiver.decodeValueOfType("@"));
		return new _NSSet(elements);
	}
};
NSSet = _NSSet = __decorate([archivedClass("NSSet")], NSSet);
let NSMutableSet = _NSMutableSet = class NSMutableSet extends NSSet {
	static initFromUnarchiver(unarchiver, archivedClass) {
		if (archivedClass.version !== 0) throw new EvalError(`Unsupported version: ${archivedClass.version}`);
		const count = unarchiver.decodeValueOfType("I");
		const elements = [];
		for (let i = 0; i < count; i++) elements.push(unarchiver.decodeValueOfType("@"));
		return new _NSMutableSet(elements);
	}
};
NSMutableSet = _NSMutableSet = __decorate([archivedClass("NSMutableSet")], NSMutableSet);
let NSDictionary = _NSDictionary = class NSDictionary extends NSObject {
	contents;
	constructor(contents) {
		super();
		this.contents = contents;
	}
	static initFromUnarchiver(unarchiver, archivedClass) {
		if (archivedClass.version !== 0) throw new EvalError(`Unsupported version: ${archivedClass.version}`);
		const count = unarchiver.decodeValueOfType("i");
		if (count < 0) throw new EvalError(`NSDictionary element count cannot be negative: ${count}`);
		const contents = /* @__PURE__ */ new Map();
		for (let i = 0; i < count; i++) {
			const key = unarchiver.decodeValueOfType("@");
			const value = unarchiver.decodeValueOfType("@");
			contents.set(key, value);
		}
		return new _NSDictionary(contents);
	}
};
NSDictionary = _NSDictionary = __decorate([archivedClass("NSDictionary")], NSDictionary);
let NSMutableDictionary = _NSMutableDictionary = class NSMutableDictionary extends NSDictionary {
	static initFromUnarchiver(unarchiver, archivedClass) {
		if (archivedClass.version !== 0) throw new EvalError(`Unsupported version: ${archivedClass.version}`);
		const count = unarchiver.decodeValueOfType("i");
		if (count < 0) throw new EvalError(`NSMutableDictionary element count cannot be negative: ${count}`);
		const contents = /* @__PURE__ */ new Map();
		for (let i = 0; i < count; i++) {
			const key = unarchiver.decodeValueOfType("@");
			const value = unarchiver.decodeValueOfType("@");
			contents.set(key, value);
		}
		return new _NSMutableDictionary(contents);
	}
};
NSMutableDictionary = _NSMutableDictionary = __decorate([archivedClass("NSMutableDictionary")], NSMutableDictionary);
//#endregion
//#region src/bplist.ts
var BPlistReader = class {
	data;
	constructor(data) {
		this.data = data;
	}
	nameMap = {
		ec: "editedContent",
		ep: "editedParts",
		rp: "retractedParts",
		euh: "editingUserHandle",
		bcg: "backwardsCompatibilityGuid",
		d: "date",
		t: "text",
		otr: "originalTextRange",
		amc: "associatedMessageContent",
		ams: "associatedMessageSummary"
	};
	process(data) {
		if (Array.isArray(data)) {
			const arr = [];
			for (const el of data) arr.push(this.process(el));
			return arr;
		} else if (data == null) return data;
		else if (data instanceof Buffer) try {
			const decoded = Unarchiver.open(data).decodeAll();
			if (decoded.length === 1) return decoded[0];
			return decoded;
		} catch {
			return data;
		}
		else if (data instanceof TypedGroup) return data;
		else if (typeof data === "object") {
			const obj = data;
			const keys = Object.keys(obj);
			if ("0" in obj && !("1" in obj)) return this.process(obj["0"]);
			if ("lo" in obj && "le" in obj && keys.length === 2) return [obj.lo, obj.le];
			for (const key of keys) if (key in this.nameMap) {
				obj[this.nameMap[key]] = obj[key];
				delete obj[key];
				obj[this.nameMap[key]] = this.process(obj[this.nameMap[key]]);
			} else obj[key] = this.process(obj[key]);
			return obj;
		}
		return data;
	}
	read() {
		const data = parseBuffer(this.data);
		this.process(data);
		return data;
	}
};
//#endregion
//#region src/archiver.ts
var TypedGroup = class {
	encodings;
	values;
	constructor(encodings, values) {
		this.encodings = encodings;
		this.values = values;
	}
};
var TypedValue = class extends TypedGroup {
	get encoding() {
		return this.encodings[0];
	}
	get value() {
		return this.values[0];
	}
	constructor(encoding, value) {
		super([encoding], [value]);
	}
};
var CArray = class {
	elements;
	constructor(elements) {
		this.elements = elements;
	}
};
var CClass = class {
	name;
	version;
	superclass;
	constructor(name, version, superclass) {
		this.name = name;
		this.version = version;
		this.superclass = superclass;
	}
};
var GenericArchivedObject = class {
	clazz;
	contents;
	constructor(clazz, contents) {
		this.clazz = clazz;
		this.contents = contents;
	}
};
var GenericStruct = class {
	name;
	fields;
	constructor(fields, name) {
		this.name = name;
		this.fields = fields;
	}
};
var NO_LOOKAHEAD = class {};
var Unarchiver = class Unarchiver {
	reader;
	binaryDecoding;
	sharedObjectTable = [];
	constructor(reader, binaryDecoding = Unarchiver.BinaryDecoding.all) {
		this.reader = reader;
		this.binaryDecoding = binaryDecoding;
	}
	static open(data, binaryDecoding = Unarchiver.BinaryDecoding.all) {
		try {
			return new Unarchiver(new TypedStreamReader(data), binaryDecoding);
		} catch (e) {
			if (e instanceof Error && e.constructor.name === "InvalidTypedStreamError") return new Unarchiver(new BPlistReader(data), binaryDecoding);
			throw e;
		}
	}
	lookupReference(ref) {
		const [refType, obj] = this.sharedObjectTable[ref.number];
		if (ref.referencedType !== refType) throw new EvalError(`Object reference type mismatch: reference should point to an object of type ${ref.referencedType}`);
		return obj;
	}
	decodeAnyUntypedValue(expectedEncoding) {
		if (this.reader instanceof BPlistReader) throw new Error("Operation not supported for datatype bplist");
		const first = this.reader.next().value;
		if (first == null || ["number", "string"].includes(typeof first) || first instanceof Buffer) return first;
		else if (first instanceof ObjectReference) return this.lookupReference(first);
		else if (first instanceof CString) {
			const str = first.contents.toString("ascii");
			this.sharedObjectTable.push([ObjectReference.Type.C_STRING, str]);
			return str;
		} else if (first instanceof Atom) return first.contents;
		else if (first instanceof Selector) return first.name;
		else if (first instanceof SingleClass) {
			const singleClasses = [first];
			let nextClassEvent = this.reader.next().value;
			while (nextClassEvent !== void 0 && !(nextClassEvent instanceof ObjectReference)) {
				if (!(nextClassEvent instanceof SingleClass)) throw new EvalError(`Expected SingleClass, ObjectReference, or undefined, not ${typeof nextClassEvent}`);
				singleClasses.push(nextClassEvent);
				nextClassEvent = this.reader.next().value;
			}
			const terminatingEvent = nextClassEvent;
			let nextSuperclass;
			if (terminatingEvent == null) nextSuperclass = void 0;
			else if (terminatingEvent instanceof ObjectReference) nextSuperclass = this.lookupReference(terminatingEvent);
			else throw new AssertionError();
			const newClasses = [];
			for (const singleClass of singleClasses.reverse()) {
				nextSuperclass = new CClass(singleClass.name, singleClass.version, nextSuperclass);
				newClasses.splice(0, 0, nextSuperclass);
			}
			for (const newClass of newClasses) this.sharedObjectTable.push([ObjectReference.Type.CLASS, newClass]);
			const result = newClasses.at(0);
			if (result == null) throw new AssertionError();
			return result;
		} else if (first instanceof BeginObject) {
			const placeholderIndex = this.sharedObjectTable.length;
			this.sharedObjectTable.push([ObjectReference.Type.OBJECT, void 0]);
			const archivedClass = this.decodeAnyUntypedValue("#");
			if (!(archivedClass instanceof CClass)) throw new EvalError(`Object class must be a CClass, not ${typeof archivedClass}`);
			const obj = archivedClassesByName.get(archivedClass.name)?.initFromUnarchiver(this, archivedClass) ?? new GenericArchivedObject(archivedClass, []);
			this.sharedObjectTable[placeholderIndex] = [ObjectReference.Type.OBJECT, obj];
			if (obj instanceof GenericArchivedObject) {
				let nextEvent = this.reader.next().value;
				while (!(nextEvent instanceof EndObject)) {
					obj.contents.push(this.decodeTypedValues(nextEvent));
					nextEvent = this.reader.next().value;
				}
			} else if (!archivedClassesByName.get(archivedClass.name)?.consumeEnd) {
				const end = this.reader.next().value;
				if (!(end instanceof EndObject)) throw new EvalError(`Expected EndObject, not ${end?.constructor.name}`);
			}
			return obj;
		} else if (first instanceof ByteArray) return new CArray([...first.data]);
		else if (first instanceof BeginArray) {
			const expectedElementEncoding = parseArrayEncoding(expectedEncoding).elementTypeEncoding;
			const decodedArray = [];
			for (let i = 0; i < first.length; i++) decodedArray.push(this.decodeAnyUntypedValue(expectedElementEncoding));
			const array = new CArray(decodedArray);
			const end = this.reader.next().value;
			if (!(end instanceof EndArray)) throw new EvalError(`Expected EndArray, not ${typeof end}`);
			return array;
		} else if (first instanceof BeginStruct) {
			const nodeStructClass = structClassesByEncoding.get(expectedEncoding);
			const fields = (nodeStructClass?.fieldEncodings ?? parseStructEncoding(expectedEncoding).fieldTypeEncodings).map((e) => this.decodeAnyUntypedValue(e));
			const end = this.reader.next().value;
			if (!(end instanceof EndStruct)) throw new EvalError(`Expected EndStruct, not ${typeof end}`);
			if (nodeStructClass == null) return new GenericStruct(fields, first.name);
			else return new nodeStructClass(fields);
		} else throw new EvalError(`Unexpected event at beginning of untyped value: ${first.constructor.name}`);
	}
	decodeTypedValues(lookahead = new NO_LOOKAHEAD()) {
		if (this.reader instanceof BPlistReader) throw new Error("Operation not supported for datatype bplist");
		const begin = lookahead instanceof NO_LOOKAHEAD ? this.reader.next().value : lookahead;
		if (!(begin instanceof BeginTypedValues)) throw new EvalError(`Expected BeginTypeValues, not ${begin?.constructor.name}`);
		const ret = begin.encodings.length === 1 ? new TypedValue(begin.encodings[0], this.decodeAnyUntypedValue(begin.encodings[0])) : new TypedGroup(begin.encodings, begin.encodings.map((e) => this.decodeAnyUntypedValue(e)));
		const end = this.reader.next().value;
		if (!(end instanceof EndTypedValues)) throw new EvalError(`Expected EndTypedValues, not ${typeof end}`);
		return ret;
	}
	decodeValuesOfTypes(typeEncodings) {
		if (typeEncodings.length === 0) throw new TypeError("Expected at least one type encoding");
		const group = this.decodeTypedValues();
		if (!allEncodingsMatchExpected(group.encodings, typeEncodings)) throw new EvalError(`Expected type encodings ${typeEncodings}, but got type encodings ${group.encodings} in stream`);
		return group.values;
	}
	decodeValueOfType(typeEncoding) {
		return this.decodeValuesOfTypes([typeEncoding])[0];
	}
	decodeArray(elementTypeEncoding, length) {
		return this.decodeValueOfType(buildArrayEncoding(length, elementTypeEncoding));
	}
	decodeDataObject() {
		const length = this.decodeValueOfType("i");
		if (typeof length !== "number" || length < 0) throw new EvalError(`Data object length cannot be negative: ${length}`);
		return this.decodeArray("c", length).elements;
	}
	decodeAll() {
		if (this.reader instanceof BPlistReader) return this.reader.read();
		else {
			const contents = [];
			while (true) {
				const lookahead = this.reader.next();
				if (lookahead.done) break;
				contents.push(this.decodeTypedValues(lookahead.value));
			}
			return contents;
		}
	}
	decodeSingleRoot() {
		const values = this.decodeAll();
		if (values.length === 0) throw new EvalError("Archive contains no values");
		else if (values.length > 1) throw new EvalError(`Archive contains ${values.length} root values (expected exactly one root value)`);
		else {
			const rootGroup = values[0];
			if (rootGroup instanceof TypedValue) return rootGroup.value;
			if (rootGroup instanceof TypedGroup) throw new EvalError(`Archive's root value is a group of ${rootGroup.values.length} values (expected exactly one root value)`);
			return rootGroup;
		}
	}
};
(function(_Unarchiver) {
	_Unarchiver.BinaryDecoding = /* @__PURE__ */ function(BinaryDecoding) {
		BinaryDecoding[BinaryDecoding["all"] = 0] = "all";
		BinaryDecoding[BinaryDecoding["decodable"] = 1] = "decodable";
		BinaryDecoding[BinaryDecoding["none"] = 2] = "none";
		return BinaryDecoding;
	}({});
})(Unarchiver || (Unarchiver = {}));
//#endregion
export { NSArray, NSAttributedString, NSData, NSDate, NSDictionary, NSMutableArray, NSMutableAttributedString, NSMutableData, NSMutableDictionary, NSMutableSet, NSMutableString, NSNumber, NSObject, NSSet, NSString, NSValue, TypedStreamReader, Unarchiver };
